ChatLog v3

krepa098 opened the pull request:
This is my work on the QGraphicsScene based chat log I've been working on during the last few months. The main goal was to reduce memory consumption and fix all the glitches of the html based chat log (smiley alignment, elided text, ...).

Other benefits:
- Custom text selection 
- Animations
- Widgets inside the chat log
- No need for base64 encoded smileys

Todo:
- [x] Check whether message history is broken or not
- [ ] Possibly change the selection algorithm to what @zetok suggested
- [x] Test with lots of messages
- [x] Implement quoting
- [x] Respect time format
- [x] fix https://gist.github.com/zetok/773876e11737200564c4
- [x] fix "Tox User" in group chats
- [x] fix anchors
- [ ] fix autoscroll (can't reproduce)
- [x] move \* to name col (done, needs color)
- [x] "load history" button
- [x] revisit selection rectangle
- [x] respect "use emoticons"
- [x] clear should reset previousId
- [x] fix crash in combination with history and clear
- [x] implement scroll while selecting
- [x] get rid of html in copied text
- [x] typing notification

I mainly created this PR to get some feedback/bug-reports. This is not ready to be merged although it should be fairly usable and stable.

Pic:
![interface3](https://cloud.githubusercontent.com/assets/4587853/5596175/7b565de4-9287-11e4-85b5-e299cde1347c.jpg)

**EDIT**: New branch at https://github.com/krepa098/qTox/commits/chatlog_v3_1 (unstashed version as requested by @dubslow )

krepa098 committed:
initial commit

krepa098 committed:
progress

krepa098 committed:
progress

krepa098 committed:
progress

krepa098 committed:
bye bye base64 emoticons

krepa098 committed:
(c)

krepa098 committed:
artwork

krepa098 committed:
progress

krepa098 committed:
filetransferwidget: first steps

krepa098 committed:
hooked up filetransferwidget

krepa098 committed:
more work on filetransferwidget

krepa098 committed:
ui tweaks

krepa098 committed:
removed ChatLine vAlignCol

krepa098 committed:
don't depend on valid cursur

krepa098 committed:
revisited selection

krepa098 committed:
changed color of selGraphItem

krepa098 committed:
margins, update selection rect during resize

krepa098 committed:
filetransferwidget: added open (browse?) button, fixed layout

krepa098 committed:
chatline: fixed bbox height calculation

krepa098 committed:
filetransferwidget.ui: reverted margins. chatlinecontentproxy: calculate
firstLineVOffset

krepa098 committed:
cleanup

krepa098 committed:
Text: use ascent rather than height

krepa098 committed:
adjusted Image, Spinner, ChatLineContentProxy vOffset

krepa098 committed:
renamed firstLineVOffset to getAscent

krepa098 committed:
Text: selection use fuzzy hit

krepa098 committed:
Text: detect anchors

krepa098 committed:
filling the gaps

krepa098 committed:
Text: toHtmlChars

krepa098 committed:
innerStyle, groupchats

krepa098 committed:
filetransferwidget layout changes

krepa098 committed:
fix selection

krepa098 committed:
chatlinecontentproxy: scaling behaviour

krepa098 committed:
filetransferwidget: use CroppingLabel

krepa098 committed:
chatlog: save chat log

krepa098 committed:
chatline: set row and col number

krepa098 committed:
chatlog: Don't accept drops

krepa098 committed:
styled chatlog, cleanup

zetok commented:
As expected, upon trying to use it I've encountered segfault.

@krepa098's branch with patches (#920, #962, #972) applied

https://gist.github.com/zetok/773876e11737200564c4

zetok commented:
Without #972 it wasn't reproducible

zetok commented:
![joins/parts](https://cloud.githubusercontent.com/assets/3148759/5596916/1b2cca66-9291-11e4-9f02-8acb5ec4aedc.png)

When `Zetok\0` peer left chat, for some reason qTox showed `:3`, as if that was the peer who left. Also, upon `Zetok\0` re-join, there's shown `Tox User` joining, instead of `Zetok\0`

zetok commented:
Actually it's reproducible without #972, crashes only when clicking on particular contacts

zetok commented:
![feedback](https://cloud.githubusercontent.com/assets/3148759/5598466/9e676c02-92ae-11e4-8c17-d202199ebe83.png)

Edit: BTW, new chat form doesn't display time in specified format – i.e., it should display also seconds on my every screenshot, but they're not displayed

apprb committed:
honouring aliases in the loaded history

Conflicts:
	src/widget/form/chatform.cpp

zetok commented:
greentexting doesn't work -.-'
![why no greentext](https://cloud.githubusercontent.com/assets/3148759/5598509/985d5e60-92af-11e4-9acb-5dbcf3adddbd.png)

krepa098 committed:
respect time format

apprb commented on src/chatlog/content/text.cpp:
```diff
@@ -0,0 +1,299 @@
+    {
+        doc = new CustomTextDocument();
+        doc->setUndoRedoEnabled(false);
+        doc->setUseDesignMetrics(true);
```
strange font rendering

![png1](https://cloud.githubusercontent.com/assets/6405341/5598827/fab15792-92e6-11e4-8224-7d97a15a4f7c.png)

Top variant: original
Bottom varian: with commented `doc->setUseDesignMetrics(true);` string

krepa098 committed:
disabled group joined/left messages

Conflicts:
	src/widget/widget.cpp

krepa098 committed:
don't use DesignMetrics

krepa098 committed:
don't set cache mode

zetok commented:
Hm, there's no way (or at least I wasn't able to spot it quickly) to load older than 1 week history.

dubslow commented:
I would very much prefer that merged version doesn't have a squashed history... and the commits added here should preferably make it back to the unsquashed version

(Still don't have a computer at the moment)

zetok commented:
![action messages](https://cloud.githubusercontent.com/assets/3148759/5600135/2cad05fe-92cc-11e4-89b2-a7c9e8c0aa0f.png)
action messages don't look good, could `*` be located right under `i`, and nick be distinguishable (bold?) from message part?
On screenshot whole `:3 urationValue of the *active* configura` is a nick, while just `asdf` is a message..

ProMcTagonist commented:
Could we get the centered boxes back instead of ( i ) messages? They look better, especially when there are a lot of them together.

dubslow commented:
@zetok How does name compare to the message in the current implementation?

zetok commented:
@krepa098 after trying to send test text sample:
![text test sample](https://cloud.githubusercontent.com/assets/3148759/5600371/8c1c19a4-92d0-11e4-8634-8eb77eba4afc.png)
I din't wait long enough to check how much more RAM qTox could actually eat, since initially after pressing "send" it used <200MB, and grew up to ~2GB

What I did:
1. Send FR to contact.
2. Before FR was accepted, I've pasted and "sent" text; qTox "hanged" right at that moment.
3. After receiving FR on other client I've accepted it and messages were received from time to time (usually few messages per minute) from hanged qTox.
**EDIT:** Received messages were not the ones that were supposed to be sent first. I.e., received messages were somewhere  from ~middle(? dunno really, just guessing) of my test sample.

# 

@dubslow it doesn't?
![muh name](https://cloud.githubusercontent.com/assets/3148759/4952764/8a2d164a-6678-11e4-99a5-6b0869659570.png)

dubslow commented:
And you'd like that behaviour to change both in the new implementation and the current one?

The messages appearing out of order is likely an issue with the DB/faux offline messaging, rather than the chat form. @apprb

zetok commented:
Why bother changing behaviour/look of old chat form that will be deprecated as soon as new one stops crashing qTox and will get missing features?

# 

Edit: not sure if I included it in my previous comments if not;
There are some problems with autoscrolling in chat; not sure what triggers it, but chat isn't always automatically scrolled down like it was in old chat form.

zetok commented:
![no, bad link, bad!](https://cloud.githubusercontent.com/assets/3148759/5601641/0a075a42-9306-11e4-9113-11f054b11b48.png)
original text:

```
1. Apply patch https://gist.github.com/zetok/a326be414c73eb64c48f
2. Recompile core and client
```

Reproducible: always

**EDIT**: another sample
![muh link](https://cloud.githubusercontent.com/assets/3148759/5602606/a8ff0cb8-9350-11e4-9faa-4335a8f65807.png)
Not a real issue, just doens't look good

krepa098 committed:
Merge branch 'master' into chatlog_v3_1

Conflicts:
	qtox.pro
	res.qrc
	src/misc/smileypack.cpp
	src/widget/chatareawidget.cpp
	src/widget/chatareawidget.h
	src/widget/form/chatform.cpp
	src/widget/form/genericchatform.cpp
	src/widget/form/genericchatform.h
	src/widget/form/settings/generalform.cpp
	src/widget/tool/chatactions/messageaction.cpp
	src/widget/tool/chatactions/systemmessageaction.cpp
	src/widget/widget.cpp

krepa098 committed:
fixup merge

krepa098 committed:
fixed anchors

krepa098 committed:
quoting support, toHtmlChars, fixes

krepa098 committed:
fixed history

zetok commented:
![distance](https://cloud.githubusercontent.com/assets/3148759/5602572/64cb2fb0-934e-11e4-838a-a4bb87065d9a.png)
↑ ~~there's quite a lot of distance between those messages, I don't know why.~~ Actually ↓
![missing msg :c](https://cloud.githubusercontent.com/assets/3148759/5602588/504f6c80-934f-11e4-9726-a7d5782b1ff5.png)

Also, I've noticed that once nick of person who has sent messages wasn't displayed in gropuchat history. After selecting text name of that peer appeared. It didn't contain any special characters AFAIK.

In the same way as nicks, some messages in groupchat history are missing.

By `groupchat history` I mean that I had to scroll way back to see them.
![no wai](https://cloud.githubusercontent.com/assets/3148759/5602583/03fcbb26-934f-11e4-99c1-0fa3adb7a0b6.png)
↑ wrong, same part of chat after selecting message ↓
![oh noes](https://cloud.githubusercontent.com/assets/3148759/5602585/0b8d454a-934f-11e4-8d26-fa3c9d50b94c.png)

krepa098 committed:
refactoring, fixes

krepa098 committed:
fix no visible line case

zetok commented:
![too muh select](https://cloud.githubusercontent.com/assets/3148759/5603100/858b2404-936b-11e4-85fe-9acdcf831241.png)
selection not always selects whole nick
qTox e8ca000379267aaf7e60c8f20e6adcbecaedd2fd with previously mentioned patches applied

zetok commented:
![emote setting](https://cloud.githubusercontent.com/assets/3148759/5603138/d3638814-936c-11e4-8b34-439eb60b075f.png)
↑ emoticons shouldn't be displayed according to settings… ↓
![nope emote](https://cloud.githubusercontent.com/assets/3148759/5603128/94cd7934-936c-11e4-9bbd-e211abc40494.png)

zetok commented:
This PR fixes #906, but at the same time invalidates #857

zetok commented:
After building patched qTox based on branch chatlog_v3_1 a0a56e32a0fcf333dbf720d46dbf4378a352e7fa so far I can no longer reproduce segfault

krepa098 committed:
check "use emoticons"

krepa098 committed:
cleanup

zetok commented:
**pause button** doesn't seem to work in file transfer widget.

Also, on a0a56e32a0fcf333dbf720d46dbf4378a352e7fa it's gotten worse with **displayed nicks**
![nick, no nicks](https://cloud.githubusercontent.com/assets/3148759/5603278/ae17dc3c-9375-11e4-932d-155012033fa8.png)

# 

also there's bug with how nick is displayed when it's long
![long nick](https://cloud.githubusercontent.com/assets/3148759/5603293/33622f28-9376-11e4-9770-67bb91cf5ef6.png)

apprb commented:
related to a0a56e3
In `GenericChatForm::addAlertMessage` and `GenericChatForm::addMessage`
`resolveToxID(..)` function should be used instead of `Core::getInstance()->getPeerName(..)` in order to get nick of every speaker in a groupchat, not only ones in your friendlist.

krepa098 committed:
ChatLine: cell spacing

krepa098 committed:
use resolveToxID

krepa098 committed:
fixed eliding (ie. invisible nicknames)

krepa098 committed:
cleanup

zetok commented:
Copying multiple messages in groupchat doesn't work well if one has a null-terminated nick and tries to copy message. Basically everything after NUL is not copied.

dubslow commented:
Gee... what a shocker... maybe we should truncate that in the UI... and thus in the chatform... well obviously we should, I wish I could get around to it

krepa098 committed:
filetransferwidget: fixed pause button, ETA, font size, resource usage

krepa098 committed:
ChatLine/ChatMessage: use shared_ptr

krepa098 committed:
cleanup

krepa098 committed:
Moved the context menu to GenericChatForm

krepa098 committed:
ChatLog: Autoscroll while selecting

zetok commented:
on a74b9ce11cea45f3ad62a0e16ede8576fb12425d loading history doesn't work

krepa098 committed:
refactoring

krepa098 committed:
ChatLog: insertChatlineAtBottom, refactoring

krepa098 committed:
history fixes

krepa098 committed:
cleanup

krepa098 committed:
changed output of ChatLog::toPlainText

zetok commented:
no idea whether latest commits were supposed to fix history loading, but it doesn't seem to work with older history on 6711fd6ee4a6fd2c40db4f89dff040b81025e3ec

What I'm doing to load all history: load history since 2013.

I have sqlite db full of messages, all weights 11.5 MB, just for testing how well new chatform is supposed to handle lots of messages.

krepa098 commented:
Does it freeze? Does it work with the current chat log?

zetok commented:
~~Doesn't freeze, just nothing happens
I have 11.5 MB big db, from which apparently I can't load history, no matter whether using new or old chat form..~~
I've found where problem lies. For some reason, qTox cannot into db loading.
How to reproduce:
1. Get some db with messages older than 1 week
2. Click load messages
3. Set year to before 2014
4. Load ← history is not being loaded.

↑ if loading history by going just by few months, then it works. 
@apprb ?

@krepa098 and yes, when I finally managed to load db, qTox "hanged". I guess it tries to load those 11.5 MB of messages, by using 100% of CPU and very slowly growing in RAM usage. Other client shows "hanged" qTox as online.

zetok commented:
So, since there's a performance issue with loading history, some perf with flamegraph: https://a.pomf.se/gtlksz.svg

suhr commented:
@krepa098 Does it do layout for every new line? And does it layout everything or some part only?

zetok commented:
![new](https://cloud.githubusercontent.com/assets/3148759/5609401/8a7dc052-9497-11e4-901c-d3243920c9ec.png)
↑ new chat form

↓ old chat form
![whoever s](https://cloud.githubusercontent.com/assets/3148759/5609404/a25f8d9a-9497-11e4-9789-3f83d16ea9be.png)

and in µTox: https://img.bi/#/XeX5Imo!KN0tzUNgdpaqAoECLlsCv94RRhLFk9nofBEauVH4

I have no idea which way is correct.

apprb commented:
@zetok can't reproduce your history loading issue

@krepa098 
Every message loaded from history marked as sent even it isn't true.
Patch to fix it:
https://gist.github.com/apprb/e4281007d924492c6731

krepa098 committed:
history: require markAsSent

krepa098 committed:
ChatLog::insertChatlineOnTop: overload taking a list

krepa098 committed:
hopefully less noticeably partial updates

krepa098 committed:
history: fixed regression

krepa098 committed:
refactoring

zetok commented:
@apprb Right. 
Between steps 

```
3. Set year to before 2014
4. Load
```

don't click on day, just set year (I did it by clicking on year and using mouse scroll) and click load.

Loading seemed to work when clicked on day.

apprb commented:
Hm, you right. Seems like it is how `QCalendarWidget` works, such actions just show corresponding month's layout.

UPDATE: fix proposed in #983

zetok commented:
I can't believe it. Amount of messages that with older chatform were causing OOM killer to be triggered now doesn't look like a problem, at least from memory standpoint on 4f01f5ec0e11db2f7c9b541fe13be09fe5dab0de

So, before loading history qTox used ~127 MB of RAM. After whole history for contact was loaded, RAM usage increased to 234 MB (tried 2 times, consistent result). Earlier same messages would eat >12GB(!) of RAM.

There are still some minor (in comparison to what was earlier) performance problems with loading history, during time it loads qTox remains hanged. Now though it takes only ~2-3 minutes to load whole history, whereas a bit earlier after >40 min history still didn't seem to be loaded.

Flamegraph of loading history: http://a.pomf.se/rwmkwu.svg

After history is loaded, there are some performance problems when scrolling through history (qTox for short moments hangs).

I haven't made flamegraph that would include performance problems with scrolling, since I have no idea to what degree it would be useful, considering that loading history itself is quite heavy task.

If such flamegraph would be useful, just say so.

Still, improvement is tremendous, thanks @krepa098, it's really awesome :3

krepa098 commented:
> If such flamegraph would be useful, just say so.

Yes, please.

> After history is loaded, there are some performance problems when scrolling through history (qTox for short moments hangs).

There is a trade-off between memory consumption and compute time.

zetok commented:
Flamegraph for scrolling after loading history: https://a.pomf.se/ihovbl.svg

zetok commented:
![nicks](https://cloud.githubusercontent.com/assets/3148759/5618168/fe550afa-950d-11e4-8c0b-3b899102435e.png)
↑ that's on version which had nicks fixed in groupchat, but it's not the latest commit; for some reason @subliun's nick after action messages wasn't displayed (I'm guessing that it's his nick)

krepa098 committed:
change previousId on action

krepa098 committed:
fixed first message not getting displayed properly

krepa098 committed:
cache innerStyle as it gets used frequently

krepa098 commented:
> Flamegraph for scrolling after loading history: https://a.pomf.se/ihovbl.svg

Thanks. I probably got rid of the Style::resolve part.

krepa098 committed:
renamed deltaRepos to deltaY

krepa098 committed:
FileTransferWidget: limit browse to pics and archives

krepa098 committed:
qrc: added missing browse_path.png

zetok commented:
Scrolling flamegraph on af575c09e6612c93c3153e1dca64704bcf1c5563: https://a.pomf.se/jlstnv.svg
Faster, but <1s "stuttering" is still present when scrolling by mouse scroll, and short hangs when using scroll bar.

krepa098 committed:
cleanup

krepa098 committed:
error symbol

krepa098 committed:
speed up ChatLine::moveBy

krepa098 committed:
[test] use std::vector rather than QVector

might be faster because std::vector doesn't do implicit sharing

krepa098 committed:
plain svg

krepa098 committed:
[test] use BspTreeIndex (might be slower?)

zetok commented:
b92d4f3d2f275b8c83765c9f1c5adca3df8c96d1
I've loaded history, scrolled for a bit and tried selecting text/messages from ~middle of history.

Selecting text is really slow. It takes several seconds to select some 'line' text.
Selecting few messages is also slow.
Deselecting(!) selected 'line' text is also really slow.
Deselecting selected messages if only a bit slow, much faster than deselecting 'line' text.

Flamegraph: https://a.pomf.se/egfmbr.svg

Edit: I think that scrolling is a bit faster than on earlier commits, but since I'm not sure to what degree placebo effect may affect me…

Edit2: All those things have been tested/measured on qTox compiled with `-O0`. Since most likely qTox will be ran almost always compiled with `-O2`, would it make significant difference? I.e., should I test its performance and provide flamegrap with `-O2`?

krepa098 commented:
Weird... BspTreeIndex is supposed to speed up operations like itemAt() but it doesn't seem to work. Could you please test 6c6d0254fc32583f2802847adcc5579d14bb5943 to see whether or not it's faster than 6c6d0254fc32583f2802847adcc5579d14bb5943?

> Edit2: All those things have been tested/measured on qTox compiled with -O0. Since most likely qTox will be ran almost always compiled with -O2, would it make significant difference? I.e., should I test its performance and provide flamegrap with -O2?

I don't know. Yes if profiling works fine with O2 enabled.

krepa098 committed:
optimized ChatLog::getContentFromPos

zetok commented:
I've checked out 6c6d0254fc32583f2802847adcc5579d14bb5943 and it seems to be faster than b92d4f3d2f275b8c83765c9f1c5adca3df8c96d1.

zetok commented:
1bf08981212bc5c265669ed9efb5b77669177926 with `-O2`
Even though it's optimized, it still seemed to be slower than 6c6d0254fc32583f2802847adcc5579d14bb5943 when scrolling.

Flamegraph: https://a.pomf.se/xqwwfz.svg

krepa098 committed:
Revert "[test] use BspTreeIndex (might be slower?)"

This reverts commit b92d4f3d2f275b8c83765c9f1c5adca3df8c96d1.

krepa098 committed:
faster scrolling, more cheating

krepa098 committed:
moved history loading to FriendWidget::setAsActiveChatroom()

krepa098 committed:
towards better scrolling

Experimental changes!
Known bug: Scrolling gets extremly slow while the cursor is inside
qTox's main window.
Cause: Unknown.

krepa098 committed:
cleanup

krepa098 committed:
Text: copy image keys

krepa098 committed:
cleanup

krepa098 committed:
Merge branch 'master' into chatlog_v3_1

Conflicts:
	src/widget/chatareawidget.cpp
	src/widget/form/chatform.h
	src/widget/form/genericchatform.cpp

krepa098 committed:
moved typing notifications to ChatLog

krepa098 committed:
tweaks, colorize the blue * in /me

krepa098 committed:
FileTransferWidget: fixed autoaccept

krepa098 committed:
cleanup

krepa098 committed:
tweaked FileTransferWidget min size

krepa098 committed:
FileTransferWidget: use baseName instead of completeBaseName + tweaked
autoAccept

zetok commented:
Problem with autoscrolling seems to appear most often when I reply after I've received message in 1 on 1 chat, or the other way around. Perhaps special symbol in my nick somehow affects it?

Also it's noticeable that after receiving message autoscroll doesn't scroll fully, leaving some bit of "space" unscrolled

krepa098 committed:
cleanup

krepa098 committed:
cleanup

krepa098 committed:
fixed chatlog not scrolling to bottom after resize

krepa098 committed:
Don't stick to bottom if the user scrolls while the worker is still
running

krepa098 committed:
comments

krepa098 committed:
More pleasant multi-line selection, refactoring

krepa098 committed:
refactoring

krepa098 committed:
tweaked ChatLog::getSelectedText

*Actions are copied as "*name action*"
*SystemInfoMessages are not copied

zetok commented:
on a0693c2fc9135d7f76aa461f4847844cc0577a63 I'm unable to save files from incoming file transfers
Edit: there's shown error window about not sufficient permissions to save file, even though I have sufficient permissions to do so..

Edit2: I think that there is a need for animated typing notification

krepa098 committed:
FileTransferWidget::acceptTransfer: Regression - check whether the path
is NOT writable

krepa098 committed:
refactoring, reverted ChatLog::getVisibleRect

krepa098 commented:
What kind of animation? Like Skype? I don't think that's entirely possibly as we get no backspace/typing events.

EDIT: Please don't edit your post if you have something important to say as I don't get notified via e-mail :)

zetok commented:
I don't know what kind of typing notification skype has, but I think that simple 3 dots appearing one after another in place of underscore would be nice (some other idea for animation?).

I don't mean that animation should _actually_ show whether someone is writing or not, but rather to show for user that qTox "works" and didn't just randomly freeze, since e.g. for me every time I see unmoving typing notification image I have a feeling that qTox just froze..

zetok commented:
![typing notification & status message](https://cloud.githubusercontent.com/assets/3148759/5785519/5e6c498e-9dcf-11e4-85dd-425d19830214.png)
Typing notification is kinda interfering with status changes

Status change:
Online → Away → Online

on b4fa8323af5e8834812e1a565067f1e87a4a7e90

krepa098 committed:
refactoring

krepa098 committed:
don't forget to reposition the typingNotification on insertChatline*

krepa098 committed:
createChatInfoMessage: increase icon size from 16 to 20 px

krepa098 commented:
>  I think that simple 3 dots appearing one after another in place of underscore would be nice

Probably a good idea. I'll have to get rid of the stylus though... 16px isn't really "a lot of space".

> Typing notification is kinda interfering with status changes

Yes, I forgot to reposition the notification.

krepa098 committed:
FileTransferWidget: calculate mean

krepa098 committed:
"%user is typing..." -> "%user ..."

krepa098 committed:
update typing notification on resize

krepa098 committed:
extended spinner class

krepa098 committed:
ChatMessage::createTypingNotification, resized images to 18x18, new
(spinning) notification icon

krepa098 committed:
updated error/info.png

dubslow committed:
minor code cleanup

zetok commented:
qTox b4fa8323af5e8834812e1a565067f1e87a4a7e90 segfaulted
https://gist.github.com/zetok/7e2fcd81f84c2a6df62e

I doubt that it's really reproducible, since it happened when I was selecting text (multiple messages), and it didn't segfault ever before when doing that (on a same version too).

Perhaps when I was selecting text there appeared new message in groupchat, could that matter?

zetok commented:
I think that spinning thing doesn't really look good. It's just my opinion, but even that unmoving notification was better.
https://a.pomf.se/usgpek.webm

ProMcTagonist commented:
> https://a.pomf.se/usgpek.webm

It's not good. It seems like it has to do with status changing because it's made out of status orbs. Completely arbitrary and confusing. If you're going to have a pretty graphic for typing indications, it should be like the one in iOS or Facebook chat: a speech bubble with an animated ellipsis.

ghost commented:
Too flashy and useless. Why not just put "xyz is typing" or ye olde three flashing dots above the input text field? Also that spinner looks off center and is bound to kick off someones OCD.

krepa098 committed:
fixed potential null deref. & memory leak

krepa098 committed:
nice!

krepa098 committed:
cleanup

suhr commented:
Why not use `. -> .. -> ... -> .` animation?

krepa098 committed:
Text: Do not rely on QTextCursor for selection

krepa098 committed:
removed selectAll()

krepa098 committed:
double-click selection

krepa098 committed:
Merge pull request #3 from dubslow/chatlog_v3_1

minor code cleanup

krepa098 commented:
I'd like to have something fluid... I do have an idea but I don't know how to implement it yet.
(No, it's not about spinning dots! :wink:  ).

krepa098 committed:
Text: extractSanitizedText, hookup double-click selection

krepa098 committed:
NotificationIcon

krepa098 committed:
renamed get/setRowIndex to get/setRow

krepa098 committed:
cleanup

krepa098 committed:
fix divide by zero

krepa098 committed:
show 2 decimals starting from MiB

krepa098 committed:
typo

zetok commented:
Hm, on 5429e50bd0c4454c2158d14014a0f87e27442208 and on some earlier commits I've encountered bug, which causes parts of messages, or whole messages to not be displayed.

I'm not exactly sure what causes it, since it's rare, but in occurrences where it happened, it seemed to have something to do with me selecting messages and scrolling(?) by selection.

I've "scrolled" with selection up and down without releasing selection.

zetok commented:
![missing message](https://cloud.githubusercontent.com/assets/3148759/5833457/4ec0406a-a14b-11e4-849a-dc18d243fe7a.png)

Edit: makes you wonder why that message wasn't displayed :D

krepa098 commented:
Fixed by 2c8e7c638d06fb464b0e3985fe46f1a5698836c3 ?

krepa098 committed:
cleanup

zetok commented:
on 5429e50bd0c4454c2158d14014a0f87e27442208 
![selection](https://cloud.githubusercontent.com/assets/3148759/5841143/75330c22-a192-11e4-9684-55cf5c1e7cac.png)
I don't really know how to reproduce, since it rarely happens.

# 

I haven't noticed "missing" messages on newer version, but I didn't use it for long…

krepa098 commented:
> I don't really know how to reproduce, since it rarely happens.

Don't worry. I know why it happens.

zetok commented:
Log saved from r-click doesn't have timestamps of any kind, I think that's a major setback.
When saving log, could it be saved in format
`[timestamp in hh:mm:ss] nick`
?

krepa098 committed:
ChatLog: busy-screen during resize

Gets rid of most ugly hacks

krepa098 committed:
Added Timestamp class, moved output generation while saving to
GenericChatForm and changed the generated layout

zetok commented:
![topic message](https://cloud.githubusercontent.com/assets/3148759/5891645/f0571718-a499-11e4-8c86-5ce87c3ad6c0.png)
notification about setting topic message is borked with certain content

thanks to @subliun

krepa098 committed:
ChatMessage::createChatInfoMessage: use toHtmlChars

krepa098 committed:
use html <center> instead of ColumnFormat::Center

krepa098 committed:
[HiDPI] ChatLog: replaced all .png by .svg, FTW custom paintEvent
This does not enable HiDPI support!

issue #975

krepa098 committed:
cleanup previous commit

krepa098 committed:
specify background brush

krepa098 committed:
reposition busy notification *after* resize

krepa098 committed:
fix regression: chatlog didn't scroll to bottom on startup

krepa098 committed:
ChatLog::selectAll, refactoring of actions (issue #808)

krepa098 committed:
make use of EmojiFontPointSize (fix #1026)

krepa098 committed:
fix showEvent calling centerOn

krepa098 committed:
FileTransferWidget: visual tweaks

krepa098 committed:
Moved status messages back to the left

krepa098 committed:
use QSize instead of QSizeF, cleanup

zetok commented:
can 882d19d311ee6bf85d252181506fbb5880571138 be reverted? `…` isn't really informative..

# 

on cfe4458d37d7efac51046d88ff3ba41e49c1b8ed:
![no typing notification](https://cloud.githubusercontent.com/assets/3148759/5936775/ac5bab26-a6df-11e4-967c-834cbd77c62d.png)
I wrote to offline `:3` from  `:3`

Not sure if that's the way it's supposed to be, there's no faux offline messaging notification nor date.

Additionally to that, there's **huge** increase in CPU usage (up to 100%) when text that was supposed to be shown as faux offline message is visible: https://a.pomf.se/upxrow.svg

krepa098 commented:
> can 882d19d be reverted? … isn't really informative..

Steam does it too. The "... is typing" part seemed rather redundant to me.

> I wrote to offline :3 from :3

You're not supposed to be able to add yourself :). 

> Not sure if that's the way it's supposed to be, there's no faux offline messaging notification nor date.

Not sure about this... does it also happen using the current chatlog?

> Additionally to that, there's huge increase in CPU usage (up to 100%) when text that was supposed to be shown as faux offline message is visible: https://a.pomf.se/upxrow.svg

This only applies to the case above, right?

zetok commented:
> > I wrote to offline :3 from :3
> 
> You're not supposed to be able to add yourself :). 

um, no, those are 2 separate profiles/IDs, just use same name

> > Additionally to that, there's huge increase in CPU usage (up to 100%) when text that was supposed to be shown as faux offline message is visible: https://a.pomf.se/upxrow.svg
> 
> This only applies to the case above, right?

~Yes. Beside that, I think that there's recently been some slight performance regression to chatform_3_1 with loading history from DB (takes more time compared to what I remember). I'll see about making flamegraph of it when I'll have time.

> Not sure about this... does it also happen using the current chatlog?

Sadly, amount of messages I load causes old chatform to be killed by OOM killer after consuming 12GB of memory, so I'm unable to test it with it…

zetok commented:
![missing notification images](https://cloud.githubusercontent.com/assets/3148759/5987767/27a1ee0a-a936-11e4-88c0-62fca54a79e4.png)

Notification about topic change is missing an indicator

zetok commented:
copying from groupchat doesn't work. no matter to which clipboard on Linux; seems like regression of some kind

krepa098 committed:
set ChatLine::isVisible to true

krepa098 committed:
Speed-up rendering

krepa098 committed:
ChatLineContent: removed boundingSceneRect()

krepa098 committed:
renamed ChatLine::boundingSceneRect to ChatLine::sceneBoundingRect

krepa098 committed:
comparators should be public

krepa098 committed:
change ChatLine::isVisible to false, rename Text::isVisible to
Text::keepInMemory

krepa098 committed:
tweaks

krepa098 committed:
fix a render artifact

krepa098 committed:
revert to "[user] is typing"

krepa098 commented:
> Notification about topic change is missing an indicator
> copying from groupchat doesn't work. no matter to which clipboard on Linux; seems like regression of some kind

![bildschirmfoto4](https://cloud.githubusercontent.com/assets/4587853/5989221/0040d2dc-a980-11e4-9166-7a4a86ee359e.png)

Seems to work fine for me. Did you try a fresh clone, make?

> Additionally to that, there's huge increase in CPU usage (up to 100%) when text that was supposed to be shown as faux offline message is visible: https://a.pomf.se/upxrow.svg

This should be fixed now.

zetok commented:
> Seems to work fine for me. Did you try a fresh clone, make?

Portage makes sure that clone is fresh, and always makes a clean build.

Issue seems to be reproducible on 783caf932c66e685c5053bef28fbf9c1baae3bc6

dubslow commented:
@krepa098 Would you be willing to force push the other branch to here? It would make following the conversation rather easier lol

krepa098 committed:
optimizations and tweaks

krepa098 committed:
Merge branch 'master' into chatlog_v3_1

Conflicts:
	qtox.pro
	res.qrc
	src/widget/form/chatform.cpp
	src/widget/form/chatform.h
	src/widget/form/genericchatform.cpp
	src/widget/tool/chatactions/chataction.h
	src/widget/tool/chatactions/messageaction.cpp
	src/widget/widget.cpp
	ui/chatArea/innerStyle.css

krepa098 committed:
fixup! Merge branch 'master' into chatlog_v3_1

krepa098 committed:
tweaked ChatLog::getSelectedText

krepa098 committed:
selection rect: change color on focus-in/out, action cleanup

krepa098 committed:
change color for precise selection on focus-in/out

krepa098 committed:
fixup! Merge branch 'master' into chatlog_v3_1

krepa098 committed:
use QPainter::setClipRect

krepa098 committed:
ChatLog: don't start the resize worker if the width doesn't change

zetok commented:
on 39b2771b8dfb70c3e7b5438f8c00285b60f93ab8

![muh link?](https://cloud.githubusercontent.com/assets/3148759/6028484/de644004-abe1-11e4-900a-d2fec2845f77.png)

krepa098 committed:
set the correct wrap mode

zetok commented:
issue with word wrapping affects all text (e.g. topic changes), not just messages

zetok commented:
some strange crash when closing qTox: https://gist.github.com/zetok/7ba7baa413cbbe5ca5a9

krepa098 committed:
fix crash on exit

zetok commented:
on ~~39b2771b8dfb70c3e7b5438f8c00285b60f93ab8~~ EDIT: it's actually on 09291bad52fb84fd716638fccdfda5a429562589 , I've copypasted wrong hash

Pending file transfer:
![button pls](https://cloud.githubusercontent.com/assets/3148759/6039668/22aa1732-ac61-11e4-8c0b-b32ee006c6cc.png)
- no buttons
- speed shouldn't be displayed
- no need for progress bar(?)

![selection pls](https://cloud.githubusercontent.com/assets/3148759/6039683/493f88dc-ac61-11e4-91e9-29ee6c90fa89.png)

zetok commented:
![highlight](https://cloud.githubusercontent.com/assets/3148759/6040051/5c652760-ac65-11e4-8bb2-097fce94a28d.png)

krepa098 committed:
add svg to CONFIG (link against the svg module)

krepa098 commented:
Seems like the SVGs don't show up. Make sure you've got qt5-svg installed (or whatever it's called on gentoo) and retry on 038a7fc.

krepa098 committed:
fix bbox calculation

krepa098 committed:
cleanup

krepa098 committed:
FTW: circular buttons [feedback wanted]

krepa098 committed:
FTW: color animation

zetok commented:
oh, so this PR introduces dependency on qtsvg to work properly? If so, then it should be reflected in `INSTALL.md` .
After installing qtsvg:5 ~~it seems to work properly~~ EDIT: icons show up.

zetok commented:
![wait for it..](https://cloud.githubusercontent.com/assets/3148759/6045975/0b1870de-ac96-11e4-8960-189b2d8fe865.png)
![dun](https://cloud.githubusercontent.com/assets/3148759/6045991/28639bbe-ac96-11e4-81c7-fd101a9d5916.png)
![pls accept](https://cloud.githubusercontent.com/assets/3148759/6046132/61cea258-ac97-11e4-9f27-7debf1a2a5d6.png)

Looks nice (?)
And animation when file transfer finishes is awesome.

suhr commented:
It's somewhat out of canon.
cast @ItsDuke

krepa098 commented:
> oh, so this PR introduces dependency on qtsvg to work properly? 

Ya, I think SVGs are the best way to tackle HiDPI support as they get rid of these `_x2.png` files.

calimeroteknik commented:
Round buttons are inconsistent with the style of the message box, so I suggest giving up on these.
![qtoxmsg](https://cloud.githubusercontent.com/assets/7931883/6046387/fceeb6ce-aca1-11e4-8ba5-0047e3a7a18d.png)

Also, the spinners don't add functionality, there already was a red mark. Is it really necessary to grab the user's attention? That's what animated stuff does.

zetok commented:
@calimeroteknik yes, spinners are necessary. They're the best: https://a.pomf.se/xanbyd.webm

dubslow commented:
Can the spinner appear next to, but not in place of, the timestamp? Just testing this for the first time since there's only one major outstanding branch now :)

ProMcTagonist commented:
Is the timestamp the time the message was created, the time it was successfully sent, or the time it was received?

calimeroteknik commented:
I tested this out earlier (5bd3cd0), and the answer was "the time it was created".

zetok commented:
on 5e4e56778ba93e2048925a5b21e3d7a694b6ad4f
There's an unnecessary pop-up window about not sufficient permissions to save file.

Steps to reproduce:
1. "send" file to qTox
2. click to save file in qTox
3. in file picker click "cancel"
4. unnecessary warning about permissions shows up

zetok commented:
on 09291bad52fb84fd716638fccdfda5a429562589
![duplicated message](https://cloud.githubusercontent.com/assets/3148759/6055919/5124b1e0-ad00-11e4-943f-e044473337a9.png)

For some reason messages were displayed 2 times.
(don't worry about missing .svg, it's an instance since before I've installed qtsvg:5 )

krepa098 commented:
> Is the timestamp the time the message was created, the time it was successfully sent, or the time it was received?

It _should_ be the time they were successfully sent.

> Can the spinner appear next to, but not in place of, the timestamp?

Well, no (technically, yes :p). Imo it doesn't make sense to show a timestamp on an undelivered message. If you sent a message on 11:00 and your friend received it on 11:05 you probably want it to display 11:05.

> For some reason messages were displayed 2 times.

Related to #1089?

> Also, the spinners don't add functionality

Sure. They tell you "it's being worked on" whereas a red line reads more like an error. I'll make them turn slower if that helps?

EDIT:

> I tested this out earlier (5bd3cd0), and the answer was "the time it was created".

Works fine for me?
![bildschirmfoto6](https://cloud.githubusercontent.com/assets/4587853/6057585/d652a3e4-ad1f-11e4-908e-9dfed8a5548b.png)

krepa098 committed:
FTW: don't display "Location not writable" if path is empty

krepa098 committed:
Spinner: lowered rotational speed

zetok commented:
@krepa098 no, spinners were perfect. There's no need to change them (haven't seen slower spinners yet)

zetok commented:
> > For some reason messages were displayed 2 times.
> 
> Related to #1089?

No. Look at timestamps

calimeroteknik commented:
> They [spinners] tell you "it's being worked on" whereas a red line reads more like an error. I'll make them turn slower if that helps?

I finally got your idea! And so I have a suggestion.
Usually, a (fast) spinner means "I'm working on it but the machine/network is slow".
It's not the case here, I think? More like "idly waiting for your contact to be online".

So yeah, I believe that making them slower is good.
I tried to look for alternatives but have nothing else to suggest, sorry. And zetok likes spinners… :)

zetok commented:
@calimeroteknik I'm sorry to say this, but please stop with this stupid bullshit about slow spinners. Now they look ugly, retarded, and qTox's "speed" decreased dramatically, to the point where it no longer feels "cool" to use it.

> Usually, a (fast) spinner means "I'm working on it but the machine/network is slow".

Bullshit. Fast spinner means that your machine is fast.

If all other programs have fast spinners (chrome, ff), and qTox doesn't have them, it looks simply as if it was slowing whole system down, thus creating an impression of bloated software that cannot even deliver messages properly.

After all, everyone knows that the faster spinners spin, the faster is your box.

EDIT:
https://a.pomf.se/nccnbw.webm _argh, my eyez!_

suhr commented:
6.5 and 6.6 are really good values for speed. Consider using 6.6 for the spinners speed.

suhr commented:
Ctrl+Insert doesn't work for copying text, neither works select to copy & middle click to paste (often used feature in linux).

dubslow commented:
I rather disagree with a slow spinner, I agree with what zetok said.

compliant commented:
Actually, a spinner is not adapted for indicating a pending message.

A spinner indicates that a background action is currently being performed ([see wikipedia](http://en.wikipedia.org/wiki/Throbber)). There is no action performed when a message is pending: the application is idle, waiting for the recipient to connect.

A better indicator would be a static _status_ indicator. Some examples:
- the red bar (change its color if it seems to mean "error");
- a static icon near the messages (a colored dot, an arrow, an elipsis "…", or a clock for "delayed");
- maybe an horizontal bar at the last position the receiver was seen online, all messages under this bar being "pending".

krepa098 committed:
faster, synchronized spinners

suhr commented:
mb different indication should be used for “user looks online, but message is not delivered yet” and “user is offline and message will be sended later”?

krepa098 committed:
cleanup

krepa098 committed:
Spinner: fade in

krepa098 committed:
cleanup

calimeroteknik commented:
Very true. A spinner would definitely be adequate when trying to deliver the message to an online user, but when the message is stored to be sent later, some static icon is preferrable. (sandglass, clock, suspension dots, offline user?)

krepa098 commented:
> mb different indication should be used for “user looks online, but message is not delivered yet” and “user is offline and message will be sended later”?
> 
> Very true. A spinner would definitely be adequate when trying to deliver the message to an online user, but when the message is stored to be sent later, some static icon is preferrable

Yes, that's pretty reasonable.

> sandglass, clock, suspension dots, offline user?

... envelope?

Anyway, I'm not going to add this now as it's not really that important and I preferably want this to get merged sooner than later.

> No. Look at timestamps

@zetok did you manage to reproduce it?

zetok commented:
@krepa098 no, after qTox's restart it was shown only once, as it should be

dubslow commented:
I also would like to get this merged as quickly as possible, even with some bugs remaining, because it's a major improvement over the current state of things.

On the other hand, discussion here has mostly devolved to nitpicks, rather than major issues that render merging a bad idea at the moment.

@krepa098 Can you enable issues in your repository, so that we may get a clear listing of the deal-breaking problems? Once the major problems are fixed I would definitely love to merge this.

krepa098 committed:
FTW: Replaced QPropertyAnimation by QVariantAnimation

krepa098 committed:
FTW: Use KeepAspectRatioByExpanding

krepa098 committed:
FTW: Don't fire an animation if start- and endcolor are the same

krepa098 committed:
FTW: Slightly increased height

krepa098 committed:
fix actions not triggering reliably + ctrl+insert shortcut on linux

krepa098 committed:
do not change clipboard content if selected text is null

krepa098 committed:
Merge branch 'master' into chatlog_v3_1

krepa098 committed:
Merge branch 'master' into chatlog_v3_1

krepa098 committed:
Text: documentLayout()->update()

krepa098 committed:
add timestamp to copied text (issue #4)

krepa098 committed:
clear previous id on action (issue #7)

krepa098 committed:
Fix ownership problem (issue #8)

krepa098 committed:
add copy to selection buffer (middle mouse-button issue #6)

krepa098 committed:
copy: handle pending messages

krepa098 committed:
tweaked layout calculation

krepa098 committed:
ownership (again)

krepa098 committed:
refactoring

krepa098 committed:
revised issue #6

zetok commented:
Well, as far as I am concerned, chatlog v3 is in a shape to be merged, [few remaining issues](https://github.com/krepa098/qTox/issues) don't look like merge blocker, and are something that can be ~easily (am I right?) fixed later on.

(tl;dr merge! - apparently that segfault is master tux3/qTox problem)

dubslow commented:
@tux3 @krepa098 

I'd like to pin down the segfault before merging, to be sure that it is actually a problem on master (and not on this). That'll have to wait until tomorrow though, giving time for the others to weigh in.

zetok commented:
Hm, I've encountered performance problem when sending **large** amounts of text (splitting); qTox is not responding, uses 100% CPU and mem usage increased from ~200 MB to >600MB before I've killed it (didn't increase rapidly). Not a merge blocker IMHO (considering amount of text), but…

Flamegraph: https://a.pomf.se/toktav.svg

suhr commented:
FIX: Underline links only on hover, change cursor on them.

krepa098 committed:
fix issue #10

krepa098 committed:
copy/paste changed format (issue #11)

krepa098 committed:
Text anchors: use PointingHandCursor

ItsDuke commented:
@suhr @zetok sorry about being a little late with this, but you're correct. For boxes like that, it's best to refer back to the original mockup:

![capture](https://cloud.githubusercontent.com/assets/7714365/6154362/2b061104-b1f9-11e4-9b2c-172e1f61f46b.PNG)

Obviously I don't know how close you'd like to get to it, but just thought I'd post it just to give an idea.

calimeroteknik commented:
+1 Yes, I pointed out earlier that the change on the buttons of the download box was inconsistent with the rest of the UI (typing field), and seems to differ from he UI guidelines.
It would be best to revert to the appearance of the above screenshot.

ProMcTagonist commented:
+1 @ItsDuke

krepa098 committed:
fix crash in Text::hoverMoveEvent

krepa098 committed:
reverse history entries as they are no longer reversed in
ChatLog::insertChatlineOnTop

krepa098 committed:
do not trigger repaint if invisible

dubslow commented:
@krepa098 Could you please restore the file transfer widget to match the mockup, as requested?

krepa098 committed:
mockup (WIP)

krepa098 committed:
Revert "FTW: circular buttons [feedback wanted]"

This reverts commit 21eb9172d4d6bbd1cabf539eacfc6327569c6e02.

krepa098 commented:
Sure. May take some time, though.

suhr commented:
BUG: cropped _posted_ emoji in new chatform when you change emoticon size from small to big
How to reproduce:
- Post an emoticon in the chat
- Change emoticon size in the settings to a bigger one
- Return to the chat

suhr commented:
Triple click doesn't select the whole line.

krepa098 committed:
finished work on FTW

hopefully

krepa098 commented:
![bildschirmfoto9](https://cloud.githubusercontent.com/assets/4587853/6199424/38e21ccc-b443-11e4-9051-025e076ac569.png)

That's how it looks now. I'll remove the 0kiB/s from paused and unaccepted file transfers.

krepa098 committed:
FTW: 1px spacing instead of 2px

krepa098 committed:
display "unconfirmed" on new file transfer requests

calimeroteknik commented:
@krepa098 nice!
Hum, why 1 pixel of spacing instead of 2? The textbox to write messages has 2px of spacing so why not do the same?

I have an issue with the text color (dark Qt theme), on this pull request's branch:
![qtoxwhite](https://cloud.githubusercontent.com/assets/7931883/6200306/1569ea20-b472-11e4-8d98-47e8d1bbf9e4.png)
On a broader debate, I'm not very happy about qTox's white background with my dark theme, but well.
At least keeping the contrast intact would be necessary!

Lastly, we reached earlier the conclusion of a static icon, for messages awaiting the other person to connect. You said an envelope or something, maybe something like ![mail-pending](https://cloud.githubusercontent.com/assets/7931883/6200539/e747e240-b47c-11e4-8f9c-bea5163a8d57.png) ? [svg](http://sprunge.us/KiKX) (just a mockup)

suhr commented:
> dark theme

Well, time to create an issue: https://github.com/ItsDuke/Tox-UI/issues/12

dubslow commented:
Major problem. The merge has been force pushed back. It increases the Qt requirements to 5.3 (QMarginsF), and that's a massive massive change that I'm not entirely sure the build system can handle. It was already a huge pain in the butt to get 5.2 working...

@krepa098 Is there an easy way for you to reduce the requirements to 5.2?

ProMcTagonist commented:
What state causes the yellow buttons? @krepa098 

@calimeroteknik tacking an icon onto every line that's awaiting sending sounds really cluttered and unnecessary. I'm in favor of the red bar, maybe with color adjustments.

calimeroteknik commented:
@ProMcTagonist currently (in this pull request) it's spinners, which is even worse for pending messages. I was also happy with the current red bar.

dubslow commented:
I like the moving icon, and I'm reasonably certain zetok does as well.

krepa098 commented:
> Major problem. The merge has been force pushed back. It increases the Qt requirements to 5.3 (QMarginsF), and that's a massive massive change that I'm not entirely sure the build system can handle. It was already a huge pain in the butt to get 5.2 working...
> Is there an easy way for you to reduce the requirements to 5.2?

I can use QMargins instead. I wasn't aware of QMarginsF being part of Qt 5.3. But keep in mind that HiDPI support requires a bump to 5.4 as well.

> What state causes the yellow buttons?

It's a file transfer that hasn't been accepted, yet.

> Hum, why 1 pixel of spacing instead of 2? The textbox to write messages has 2px of spacing so why not do the same?

It seems like the spacing is only 1px in the mockup. So I guess the textbox is wrong :P

> Lastly, we reached earlier the conclusion of a static icon, for messages awaiting the other person to connect. You said an envelope or something, maybe something like mail-pending ? svg (just a mockup)

Yes, I remember that. It's on my TODO list but it's not going to be part of this PR. (I like the mockup btw)

zetok commented:
@krepa098 please don't change faux offline message indicator to something static. Current one is really good.

suhr commented:
> different indication should be used for “user looks online, but message is not delivered yet” and “user is offline and message will be sended later”

Just use spinners only for the first.

dubslow commented:
When Debian Jessie becomes stable, we can reexamine Qt versions. (There may still be issues on the Fedora family of things, but Wheezy is a pain in the butt and I don't want to change anything until we can retire Wheezy.)

Edit: If you can, I'd like to get this down to 5.2 and merged as soon as possible -- if you're around, an hour or less.

dubslow commented:
@krepa098 Since you said they're replaceable, I took a whack at fixing it myself, and Jenkins seemed to compile okay. Having said that, can you look over the commits in master to be sure?

krepa098 commented:
@dubslow I think a did all the changes I wanted to do so I think you can attempt another merge :)

krepa098 commented:
@dubslow Yes that's fine. But there are a bunch of fixes left in my branch (hopefully without regressions).

calimeroteknik commented:
@krepa098 I just saw that in unicode there are `📨 📩 ✉ 🕒`, this may be of use.
As ProMcTagonist noted, it looks cluttered to have several, since messages will get delivered in order anyway, so one may suffice. On the other hand, with a lot of messages pending, it might be out of sight.
The original red line was indeed nice for pending messages… (spinners being totally OK for messages not yet ACKed)

What could be done about the font color being very like the background, when using a dark theme?
Maybe the fix can come from Tox accepting a darker background for the message pane, but in the meantime it's really unreadable (see earlier screenshot).

ProMcTagonist commented:
I like spinners for pending, red bar for offline friend (with a firstrun/tooltip message about what's happening).

One thing - spinners should **not** appear immediately. It makes it seem really laggy when every single message has a brief pending period. Maybe only do the spinner after a second.

calimeroteknik commented:
> Maybe only do the spinner after a second.

Did you try this out? It already fades in with a delay.
Anyway it's cool, since it tells you when the message was acknowledged by your contact's client.

ProMcTagonist commented:
Looks like I was just testing on a slow bot. It's better everywhere else.
